### PR TITLE
S6 t2/create frontend displaying upcoming meetings

### PR DESCRIPTION
**Pull request by catai9**
**Scrum Board Ticket:** S6 T2: Create the front end interface for displaying all upcoming appointments on the dashboard.

**Story:** Story 6 Meeting/Appointment Descriptions

**Description**: Added an AppointmentList component to display the upcoming appointments on the Home page and also installed dependencies to run the unit tests and sass. This PR also involves updating the navigation bar to display softer colours for worker and student users (as per the TA feedback).

**Testing Criteria:** Tested code by running locally and ensuring that the Home page displayed as expected for both the worker and student users. I also ran the automated tests codes and ensured that there was 100% file coverage (and passing tests) for all the server tests. I also ran the unit tests in the front end and ensured 100% file coverage (and passing tests) for the AppointmentList component. The coverage for the other tests are in scope for another ticket and not set up yet (hence why less than 100% file coverage for those files). 

**How Reviewers can Test Code:**
1. Run the code locally and ensure no complications when starting up (run the command `npm run-script start:dev` in the root folder OR run the command `npm start` if your node and npm versions are above 12.10.0 for node nad 6.10.3 for npm).
a. Use the credentials `username: johndoe@gmail.com and password: a1234` to view the upcoming appointments. The following window should appear: 
image
b. Use the credentials `username: joshuabrooks@gmail.com and password: j1234` to view the upcoming appointments. The following window should appear:  
image

2. Run the automated server tests locally and ensure all tests pass and have 100% file coverage. 
a. Navigate inside the `server` folder and run `npm test`. 
b. Navigate inside the `react-ui` folder and run `npm test`. All tests should pass and the AppointmentList component should have 100% file coverage (the rest are not set up yet and are in the scope for another ticket).

### CODE REVIEW

File: react-ui/src/App.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ class App extends Component {
         <Route path="/login" component={LogIn}>
           {/* <LogIn /> */}
         </Route>
-        <Route exact path="/dashboard/:email/:type/:name" component={Dashboard}>
+        <Route exact path="/dashboard/:email/:userType/:firstName/:personId/:accessToken" component={Dashboard}>
```
Comment (catai9) [inline react-ui/src/App.js:31]: Please note that additional props were needed from the login page. Although this method is not a secure way of passing sensitive parameters such as accessToken, adjustments such as implementing a redux store or leveraging another method is out of scope for this ticket and will be considered for a future sprint.

File: react-ui/src/__snapshots__/App.test.js.snap [hunk 2]
```diff
@@ -0,0 +1,119 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (catai9) [inline react-ui/src/__snapshots__/App.test.js.snap:1]: Please note that snapshot code can be effectively ignored as they are created automatically and used only for testing comparisons.

File: react-ui/src/App.test.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import React from "react";
```
Comment (catai9) [inline react-ui/src/App.test.js:1]: Please note that this PR only sets up the initial test for the App component. Additional tests are in scope for another ticket and out of scope for this ticket.

File: react-ui/src/constants/userTypes.json [hunk 4]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (catai9) [inline react-ui/src/constants/userTypes.json:1]: A file was created to standardize the userTypes across the front end components. Unfortunately, this constants file cannot be in the root folder (for sharing across the server and react-ui codebase) due to limitations with React.

File: server/package.json [hunk 5]
```diff
@@ -11,8 +11,6 @@
     "dotenv": "^4.0.0",
     "express": "^4.16.2",
     "joi": "^17.2.1",
-    "lodash": "^4.17.20",
```
Comment (catai9) [inline server/package.json:14]: These dependencies were moved to the main folder to reduce unnecessary repetition as they are used in both the react-ui and server codebases.

File: server/models/data/Student.js [hunk 6]
```diff
@@ -7,7 +7,7 @@ class Student {
         this.firstName = studentData.first_name;
         this.lastName = studentData.last_name;
         this.email = studentData.email;
-        this.type = UserTypes.student;
```
Comment (catai9) [inline server/models/data/Student.js:10]: Please note that this change was due to the conflicting issue with the worker type attribute.

File: react-ui/src/components/Layouts/Profile.js [hunk 7]
```diff
@@ -1,7 +1,7 @@
 import React, { Component, Fragment } from "react";
 import Title from "../Title"
 
-class Page2 extends Component {
```
Comment (catai9) [inline react-ui/src/components/Layouts/Profile.js:4]: This file was renamed to increase clarity and avoid user confusion (as mentioned in the TA demo) but no actual work was done for this page for this PR (out of scope).

File: react-ui/src/components/Layouts/Home.js [hunk 8]
```diff
@@ -1,13 +1,16 @@
 import React, { Component, Fragment } from "react";
+
 import Title from "../Title"
+import AppointmentList from "../AppointmentList/AppointmentList";
 
-class IndexDashboard extends Component {
+class Home extends Component {
```
Comment (catai9) [inline react-ui/src/components/Layouts/Home.js:6]: Renamed this page to Home to increase clarity as IndexDashboard may be confusing.

File: react-ui/src/components/Dashboard/Dashboard.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
+import React, { Component, Fragment, Router } from "react";
```
Comment (catai9) [inline react-ui/src/components/Dashboard/Dashboard.js:1]: Please note that I moved Dashboard to its own folder to increase clarity as it is different from the typical Layouts and is used to control the navigation across all components (after login).

File: react-ui/src/App.js (hunk 1 above)
Comment (sehersarin) [reply react-ui/src/App.js:31]: This looks good, thank you for the reasoning.

**Comment by sehersarin** [comment]
Hi Amy, took a look through your PR and tried running the code however, I am getting this error - do you know why this could be? (I ran the npm install command and then npm start in the react-ui folder.)

Screen Shot 2020-11-01 at 1 57 33 PM

**Comment by sehersarin** [comment]
Additionally, I see that there are 3 failing checks for CI - would you please be able to take a look.

File: react-ui/src/components/Dashboard/Dashboard.js (hunk 9 above)
Comment (sehersarin) [reply react-ui/src/components/Dashboard/Dashboard.js:1]: That makes sense!

File: react-ui/src/components/Layouts/Home.js (hunk 8 above)
Comment (sehersarin) [reply react-ui/src/components/Layouts/Home.js:6]: Sounds good to me.

File: react-ui/src/components/Layouts/Profile.js (hunk 7 above)
Comment (sehersarin) [reply react-ui/src/components/Layouts/Profile.js:4]: Yes, I agree - thank you for renaming this file, will definitely help avoid any confusion.

File: react-ui/src/App.test.js (hunk 3 above)
Comment (sehersarin) [reply react-ui/src/App.test.js:1]: Great, thanks for clarifying the scope of this ticket.

File: react-ui/src/constants/userTypes.json (hunk 4 above)
Comment (sehersarin) [reply react-ui/src/constants/userTypes.json:1]: Understood - this does make sense.

**Comment by mtatran** [comment]
Screen Shot 2020-11-01 at 2 46 06 PM
 Jest tests are failing

**Comment by sehersarin** [comment]
> Hi Amy, took a look through your PR and tried running the code however, I am getting this error - do you know why this could be? (I ran the npm install command and then npm start in the react-ui folder.)
> 
> Screen Shot 2020-11-01 at 1 57 33 PM

Update: Was able to fix this by doing an npm install in the root folder.

**Comment by sehersarin** [comment]
Have verified the UI for both login credentials a) and b) and they look good to me! 

However, when I try to run test cases there are some errors or low code coverage. 
1) When running npm test in react-ui folder this is what I see (screenshot below)

Screen Shot 2020-11-01 at 2 58 44 PM


2.) When running npm test in server folder it shows 3 tests and 3 test suites fail.

Screen Shot 2020-11-01 at 2 59 54 PM

**Comment by catai9** [comment]
Hi Seher, 
Glad to hear that worked! 
1. I fixed the failing checks by deleting and reinstalling the dependencies again. 
2. I also updated the test files in the server folder to account for the change in `userType`. All server tests should be passing now. 
3. The red lines on the front end tests are not errors - they signify uncovered lines and low code coverage (which is in scope for another task). In your screenshot, you can see that the AppointmentList component contains 100% code coverage. As shown below, these lines (located at the bottom) indicate if tests pass:  
image.

**Comment by catai9** [comment]
> 
> 
> Screen Shot 2020-11-01 at 2 46 06 PM
> Jest tests are failing

Hi Melissa, you might need to do `npm install` again to make sure that all of your dependencies are up-to-date.

**Comment by sehersarin** [comment]
Hi Amy, 

Thank you for making all of the requested changes and fixing the errors for the server tests. Have verified the UI and code coverage and everything looks good to me. I'm ready to approve and merge the PR to the master branch.

**Review by sehersarin — APPROVED** [review]

**Pull request merged**